### PR TITLE
Extend the OpenConsoleFeature

### PR DESCRIPTION
**Pull request by obraliar**
Extend the OpenConsoleFeature by offering a selection between default terminal emulator and configurable command execution.

@koppor 's wish (see https://github.com/JabRef/jabref/issues/462#issuecomment-191802851) was to have a selection between own terminal emulators. This feature extension provides that.

- Add:
  - radio selection to the ExternalTab
  - program execution to JabRefDesktop
  - new JabRefPreference entries
  - UI outputs
  - localization keys

Screenshots:
oc_sc3

- [x] Change in CHANGELOG.md described
- [x] Screenshots added (for bigger UI changes)

### CODE REVIEW

**Comment by tschechlovdev** [comment]
please fix codacy

**Comment by obraliar** [comment]
Done.

File: src/main/java/net/sf/jabref/gui/preftabs/AdvancedTab.java [hunk 1]
```diff
@@ -61,6 +77,39 @@ public AdvancedTab(JabRefPreferences prefs) {
         useCaseKeeperOnSearch = new JCheckBox(Localization.lang("Add {} to specified title words on search to keep the correct case"));
         useUnitFormatterOnSearch = new JCheckBox(Localization.lang("Format units by adding non-breaking separators and keeping the correct case on search"));
 
+        defaultConsole = new JRadioButton(Localization.lang("Use default terminal emulator"));
+        specifiedConsole = new JRadioButton(Localization.lang("Specify terminal emulator:"));
+        consoleEmulatorPath = new JTextField(20);
```
Comment (boceckts) [inline src/main/java/net/sf/jabref/gui/preftabs/AdvancedTab.java]: Why did you set the columns to 20? In my opinion you can use more as it seems there is still space to the right in the preference dialog.

File: src/main/java/net/sf/jabref/gui/preftabs/AdvancedTab.java [hunk 2]
```diff
@@ -164,22 +226,57 @@ private void storeRemoteSettings() {
...
             JOptionPane.showMessageDialog(null,
                     Localization.lang("You must enter an integer value in the interval 1025-65535 in the text field for")
                             + " '" + Localization.lang("Remote server port") + '\'',
                     Localization.lang("Remote server port"), JOptionPane.ERROR_MESSAGE);
-            return false;
+            isValid = false;
         }
+
+        if (!consoleEmulatorPath.getText().trim().isEmpty()) {
+            File file = new File(consoleEmulatorPath.getText());
```
Comment (boceckts) [inline src/main/java/net/sf/jabref/gui/preftabs/AdvancedTab.java]: Please try to use java nio methods.

File: src/main/java/net/sf/jabref/gui/desktop/JabRefDesktop.java [hunk 3]
```diff
@@ -309,7 +310,13 @@ public static void openConsole(File file) throws IOException {
...
 
         String absolutePath = file.toPath().toAbsolutePath().getParent().toString();
-        NATIVE_DESKTOP.openConsole(absolutePath);
+        File console = new File(Globals.prefs.get(JabRefPreferences.CONSOLE_APPLICATION));
+
+        if (Globals.prefs.getBoolean(JabRefPreferences.USE_DEFAULT_CONSOLE_APPLICATION)) {
+            NATIVE_DESKTOP.openConsole(absolutePath);
+        } else if (console.exists() && console.isFile() && console.isAbsolute()) {
+            Runtime.getRuntime().exec(console.getAbsolutePath(), null, new File(absolutePath));
+        }
```
Comment (boceckts) [inline src/main/java/net/sf/jabref/gui/desktop/JabRefDesktop.java]: Please add a java doc comment. Also mention that this depends on the user settings.

File: src/main/java/net/sf/jabref/gui/desktop/JabRefDesktop.java [hunk 4]
```diff
@@ -309,7 +310,13 @@ public static void openConsole(File file) throws IOException {
         }
 
         String absolutePath = file.toPath().toAbsolutePath().getParent().toString();
-        NATIVE_DESKTOP.openConsole(absolutePath);
+        File console = new File(Globals.prefs.get(JabRefPreferences.CONSOLE_APPLICATION));
```
Comment (boceckts) [inline src/main/java/net/sf/jabref/gui/desktop/JabRefDesktop.java]: Please try to use java nio methods.

File: src/main/java/net/sf/jabref/gui/desktop/os/Windows.java [hunk 5]
```diff
@@ -48,8 +48,6 @@ public void openFolderAndSelectFile(String filePath) throws IOException {
 
     @Override
     public void openConsole(String absolutePath) throws IOException {
-        ProcessBuilder process = new ProcessBuilder("cmd.exe", "/c", "start");
-        process.directory(new File(absolutePath));
-        process.start();
+        Runtime.getRuntime().exec("cmd.exe /c start", null, new File(absolutePath));
```
Comment (boceckts) [inline src/main/java/net/sf/jabref/gui/desktop/os/Windows.java]: 👍 for not using `ProcessBuilder`

**Comment by boceckts** [comment]
Very nice PR, only minor comments from my side 👍

File: src/main/java/net/sf/jabref/gui/desktop/os/Windows.java (hunk 5 above)
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/gui/desktop/os/Windows.java]: Be careful with that! 
It was change to PB with intention in https://github.com/tschechlovdev/jabref/commit/73f69e6c2a6ddcaef8a14207a1967f813c3fdf24

File: src/main/java/net/sf/jabref/gui/preftabs/AdvancedTab.java (hunk 1 above)
Comment (obraliar) [inline src/main/java/net/sf/jabref/gui/preftabs/AdvancedTab.java]: Updated to 25. It should be ok right now.

File: src/main/java/net/sf/jabref/gui/desktop/JabRefDesktop.java (hunk 3 above)
Comment (obraliar) [inline src/main/java/net/sf/jabref/gui/desktop/JabRefDesktop.java]: Done.

File: src/main/java/net/sf/jabref/gui/desktop/os/Windows.java (hunk 5 above)
Comment (obraliar) [inline src/main/java/net/sf/jabref/gui/desktop/os/Windows.java]: Undone.

File: src/main/java/net/sf/jabref/gui/preftabs/AdvancedTab.java (hunk 2 above)
Comment (obraliar) [inline src/main/java/net/sf/jabref/gui/preftabs/AdvancedTab.java]: Done using `Paths`.

File: src/main/java/net/sf/jabref/gui/desktop/JabRefDesktop.java (hunk 4 above)
Comment (obraliar) [inline src/main/java/net/sf/jabref/gui/desktop/JabRefDesktop.java]: Done using `Paths`.

File: src/main/resources/l10n/JabRef_da.properties [hunk 6]
```diff
@@ -1706,3 +1706,8 @@ Online_help_forum=
 Custom=
 
 Converts_HTML_code_to_Unicode.=
+
+Open_console=
+Please_type_in_the_absolute_path_to_an_existing_terminal_emulator.=
+Specify_terminal_emulator\:=
```
Comment (koppor) [inline src/main/resources/l10n/JabRef_da.properties]: Do not include final `\:` in the language files

Comment (obraliar) [inline src/main/resources/l10n/JabRef_da.properties]: Done.

File: src/main/java/net/sf/jabref/gui/preftabs/AdvancedTab.java [hunk 7]
```diff
@@ -61,6 +79,39 @@ public AdvancedTab(JabRefPreferences prefs) {
         useCaseKeeperOnSearch = new JCheckBox(Localization.lang("Add {} to specified title words on search to keep the correct case"));
         useUnitFormatterOnSearch = new JCheckBox(Localization.lang("Format units by adding non-breaking separators and keeping the correct case on search"));
 
+        defaultConsole = new JRadioButton(Localization.lang("Use default terminal emulator"));
```
Comment (koppor) [inline src/main/java/net/sf/jabref/gui/preftabs/AdvancedTab.java]: The preferences should be in "External programs", not in "Advanced", as "Console" is an external program.

**Comment by koppor** [comment]
Locally, I got following exceptoin

```
:compileJavaC:\git-repositories\jabref\jabref\src\main\java\net\sf\jabref\BibDatabaseContext.java:1: error: An unhandled exception was thrown by the Error Prone static analysis plugin.
package net.sf.jabref;
^
     Please report this at https://github.com/google/error-prone/issues/new and include the following:

     error-prone version: 2.0.10
     Stack Trace:

```

After disabling the error prone plugin (just remove the line with `errorprone` in build.gradle), I could test it.

My preferences look like this:

grabbed_20160717-150310

Why is the space to the right not fully used?

Other categories of the preferences do use it:

grabbed_20160717-150434


Featurewise: Just using `C:\Program Files\ConEmu\ConEmu64.exe` just starts a new instance. The old instance has to be reused. Thus, my command line is `ConEmu64.exe /single /dir "c:\users"`. As a consequence, the whole parameter string should be configurable with `%DIR` as placeholder for the directory name.

File: src/main/java/net/sf/jabref/gui/preftabs/AdvancedTab.java (hunk 7 above)
Comment (obraliar) [inline src/main/java/net/sf/jabref/gui/preftabs/AdvancedTab.java]: Done. Moved to `ExternalTab`.

**Comment by obraliar** [comment]
Okay, I'm going to add a new Option called "execute". Where should I describe that `%DIR` is a placeholder? Below, as labeled text?

File: src/main/java/net/sf/jabref/gui/preftabs/ExternalTab.java [hunk 8]
```diff
@@ -113,22 +172,66 @@ public void setValues() {
...
     }
 
     @Override
     public boolean validateSettings() {
+        if (!consoleEmulatorPath.getText().trim().isEmpty()) {
+            Path path = Paths.get(consoleEmulatorPath.getText());
+
+            if (!Files.exists(path) || Files.isDirectory(path) || !path.isAbsolute()) {
+                JOptionPane.showMessageDialog(null,
+                        Localization.lang("Please type in the absolute path to an existing terminal emulator."),
```
Comment (oscargus) [inline src/main/java/net/sf/jabref/gui/preftabs/ExternalTab.java]: Maybe I shouldn't interfere at this time, but I think replacing "type in" with "enter" would make sense.

**Comment by koppor** [comment]
Labeled text sounds good.

File: src/main/java/net/sf/jabref/gui/preftabs/ExternalTab.java (hunk 8 above)
Comment (obraliar) [inline src/main/java/net/sf/jabref/gui/preftabs/ExternalTab.java]: Done.

**Comment by obraliar** [comment]
The feature is ready. Please test the feture with some terminal emulators.

_Screenshot:_
oc_sc1

**Comment by koppor** [comment]
Execute Command only is sufficient. Specify terminal emulator is not necessary IMHO.

When I press the button, I get no feedback. Typically, JabRef writes "Externel viewer called" or something like that. Here, nothing is done (and nothing happened here).

Please double check on windows whether the setting "Use default terminal emulator" still works.

There is no need to underline the note field and to put %DIR in italics. Just write it in plain text. This also makes it easier for translators.

Put `C:\Program Files\ConEmu\ConEmu64.exe /single /dir "%DIR"` as default value for "execute command" (on Windows).

Works also with paths with spaces: :+1:

File: src/main/java/net/sf/jabref/gui/desktop/JabRefDesktop.java [hunk 9]
```diff
@@ -330,10 +330,17 @@ public static void openConsole(File file) throws IOException {
             NATIVE_DESKTOP.openConsole(absolutePath);
         } else if (usingCommand) {
             String command = Globals.prefs.get(JabRefPreferences.CONSOLE_COMMAND);
+            command = command.trim();
+
             if (!command.isEmpty()) {
-                command = command.replace("%DIR", absolutePath); // replace the placeholder if used
+
```
Comment (koppor) [inline src/main/java/net/sf/jabref/gui/desktop/JabRefDesktop.java]: No need for an empty line :innocent:

File: src/main/java/net/sf/jabref/gui/preftabs/ExternalTab.java [hunk 10]
```diff
@@ -88,7 +88,7 @@ public ExternalTab(JabRefFrame frame, PreferencesDialog prefsDiag, JabRefPrefere
         JLabel commandDescription = new JLabel(
                 "<html>" +
                         Localization.lang("<u>Note</u>: Use the placeholder <i>%0</i>" +
-                                " for the location of the current database file.", "%DIR") +
```
Comment (koppor) [inline src/main/java/net/sf/jabref/gui/preftabs/ExternalTab.java]: I liked "current" more than "opened": There is always one "current" database file (or zero if nothing is opened :)), but possibly more than one opened.

File: src/main/java/net/sf/jabref/JabRefPreferences.java [hunk 11]
```diff
@@ -868,7 +868,12 @@ private JabRefPreferences() {
         defaults.put(INDENT, 4);
 
         defaults.put(USE_DEFAULT_CONSOLE_APPLICATION, Boolean.TRUE);
-        defaults.put(CONSOLE_COMMAND, "");
+
```
Comment (koppor) [inline src/main/java/net/sf/jabref/JabRefPreferences.java]: Remove empty line here

File: src/main/java/net/sf/jabref/gui/desktop/JabRefDesktop.java [hunk 12]
```diff
@@ -321,32 +318,42 @@ public static void openConsole(File file) throws IOException {
...
                     subcommands[i] = subcommands[i].replace("%DIR", absolutePath); // replace the placeholder if used
+                    commandLoggingText.append(subcommands[i]);
+                    if (i < (subcommands.length - 1)) {
+                        commandLoggingText.append(" ");
+                    }
                 }
 
-                new ProcessBuilder(subcommands).start();
+                JabRefGUI.getMainFrame().output(Localization.lang("Executing the command \"%0\"...", commandLoggingText.toString()));
+                LOGGER.info(Localization.lang("Executing the command \"%0\"...", commandLoggingText.toString()));
```
Comment (koppor) [inline src/main/java/net/sf/jabref/gui/desktop/JabRefDesktop.java]: Do not localize LOGGER messages

**Comment by obraliar** [comment]
@koppor 
> Execute Command only is sufficient. Specify terminal emulator is not necessary IMHO. - Use default terminal emulator is good. That means: Please provide two choices. The Browse button can go to execute command and replace the whole string when it is pressed.

"Specify terminal emulator" option has been removed.

> When I press the button, I get no feedback. Typically, JabRef writes "Externel viewer called" or something like that.

Done.

> Put C:\Program Files\ConEmu\ConEmu64.exe /single /dir "%DIR" as default value for "execute command" (on Windows).

Done.

**Comment by obraliar** [comment]
Here is the feature now looks like _(default setting)_:

oc_sc2

**Comment by obraliar** [comment]
Here is a new screenshot _(default setting)_:

oc_sc3

**Comment by obraliar** [comment]
Refactoring: Done.

**Comment by koppor** [comment]
LGTM :+1:

**Comment by lenhard** [comment]
I can confirm that this works with ConEmu on my machine. I have a different path to ConEmu than the default, but changing the preferences worked without an issue. Can be merged from my point of view.

**Comment by matthiasgeiger** [comment]
Good work!

**Pull request merged**